### DISCUSSION TITLE
Please explain migration from Dotlin

### DISCUSSION THREAD

**Discussion by eggnstone** [dotlin-org/dotlin#111] [Q&A] [answered]
Is Dotlin dead?
Will Dirt compile Kotlin to Dart?

**Comment by wilkomanger** [comment, marked as answer]
Hi, thanks for the question!

Dotlin is "dead", but Dirt is is the continuation of it, because both Dotlin and Dirt have the same exact goal: Making it possible to use a more beautiful programming language that targets the Dart runtime, and fully integrates into the Dart ecosystem.

As you could tell before, Dotlin itself was not "pure Kotlin", it had it's own features, like a more lenient Dart-like `const`, or simplified primitives (just `Int` and `Double`). So calling it a "Kotlin to Dart compiler" was my mistake :sweat_smile:

Dotlin did use the Kotlin compiler for the initial phase, but as Dotlin grew, the Kotlin compiler got more and more in the way. It also limited me in what features I could implement, or how well I could implement them. For example, for calling a `const` constructor directly, I had to create an annotation `@const`, to use like so `val x = @const MyClass()`. This is because of a limitation in the Kotlin compiler, it simply doesn't support keywords in front of expressions.

So I've decide to rewrite, and rename Dotlin to Dirt. Dirt is now written in Dart, and it's a fully independent compiler now, which means I have way more freedom in how the syntax will look like! The base syntax is still heavily inspired by Kotlin, but I'm also fixing inconsistencies or things that were missing from the Kotlin syntax, and adding new unique features. So Dirt is really it's own language, like Dotlin, but more independent now.

I've chosen to rename Dotlin to Dirt to reflect this, as the primary relationship with the project is Dart, not Kotlin. Not per se in language design (although of course Dirt has Dart features), but in Dart integration.

So the goal of Dotlin and Dirt is exactly the same (creating a clean language for the Dart runtime), but now I'm not limited by the Kotlin compiler anymore, and I'm more free to create the language!

Hope that helps, and I hope the project will still be useful to you!

**Comment by eggnstone** [comment]
I guess I'll have write a Kotlin-to-Dirt compiler then :D

Sucks a bit because I really wanted to make use of your Kotlin-to-Dart compiler and I can't find any other one.

**Reply by wilkomanger** [reply]
Ha! It's not impossible to make a pure Kotlin to Dart (or Dirt) compiler, but there are a lot of challenges because there are a lot of inconsistencies you're gonna have to handle. Random example is how Kotlin's `Iterable` is eager but Dart's `Iterable` is lazy. In Kotlin you have `Sequence` but if you choose to use that as a stand-in for Dart's `Iterable`, you're gonna have to handle the fact that Kotlin's `List` is not a subtype of `Sequence`, but Dart's `List` _is_ a subtype of `Iterable`. Of course there are work arounds for it like compiling type-checks specially, but it's not ideal. You're gonna have to do a lot of those kind of fixes :sweat_smile: 

Not impossible but still a challenge!

**Reply by wilkomanger** [reply]
Btw can you remind me what's your exact use-case for a pure Kotlin-to-Dart compiler? Just curious

**Comment by eggnstone** [comment]
I'm building a dart formatter, but it needs to be written in Java or Kotlin because it should run in IntelliJ/AndroidStudio. But I also want to use it as a dart pub.dev package so it can be run in a pipeline. And I don't want to write it twice, so that would have been a nice solution.

**Reply by wilkomanger** [reply]
Ah yes, I see :+1: I suppose you've considered it already, but you could call the Dart executable from the IntelliJ plugin and keep all the logic in Dart

**Reply by eggnstone** [reply]
> Ah yes, I see 👍 I suppose you've considered it already, but you could call the Dart executable from the IntelliJ plugin and keep all the logic in Dart

Haven't thought about that yet. Thanks!

**Reply by wilkomanger** [reply]
No problem, good luck!
